sift/src/main/java/siftscience/android/Uploader.java

```diff
@@ -1,357 +1,213 @@
+    }
 
-        void reset() {
-            request = null;
-            backoff = initialBackoff;
-            nextUploadTime = 0;
-            numRejects = 0;
+    public void upload(List<MobileEventJson> batch) {
+        // Kick-off the first upload
+        try {
+            byte[] requestBody = buildRequest(batch);
+            if (requestBody != null) {
+                Log.d(TAG, String.format("Uploading batch of size %d", batch.size()));
+                this.doUpload(requestBody, MAX_RETRIES);
+            }
+        } catch (IOException e) {
+            Log.e(TAG, "Encountered IOException in upload", e);
         }
     }
 
-    /**
-     * Wrapper of a FIFO of events (each group of events is called a
-     * batch) to provide atomic operations.
-     */
-    private static class Batches {
-        @SerializedName(value="batches")
-        private final List<List<MobileEventJson>> batches;
-
-        Batches() {
-            batches = new LinkedList<>();
+    private void doUpload(byte[] requestBody, int retriesRemaining) {
+        if (retriesRemaining == 0) {
+            return;
         }
 
-        String archive() throws JsonParseException {
-            return Sift.GSON.toJson(this);
-        }
+        this.taskManager.schedule(
+                new UploadTask(this, requestBody, retriesRemaining),
+                (long) (Math.pow(MAX_RETRIES - retriesRemaining, BACKOFF_EXPONENT) * BACKOFF_MULTIPLIER),
+                BACKOFF_UNIT
+        );
+    }
 
-        /** Append a batch to the end of FIFO. */
-        synchronized void append(List<MobileEventJson> batch) {
-            batches.add(batch);
+    /** Builds an HTTP request for the specified event batch */
+    @Nullable
+    private byte[] buildRequest(List<MobileEventJson> batch) throws IOException {
+        if (batch.isEmpty()) {
+            return null;
         }
 
-        /** Peek the first batch. */
-        synchronized List<MobileEventJson> peek() {
-            return batches.isEmpty() ? null : batches.get(0);
-        }
+        Sift.Config config = configProvider.getConfig();
 
-        /** Remove the first batch. */
-        synchronized void pop() {
-            if (!batches.isEmpty()) {
-                batches.remove(0);
-            }
+        if (config == null) {
+            Log.d(TAG, "Missing Sift.Config object");
+            return null;
         }
-    }
 
-    private final ScheduledExecutorService executor;
-    private final State state;
-    private final Future unarchiveFuture;
+        if (config.accountId == null || config.beaconKey == null || config.serverUrlFormat == null) {
+            Log.d(TAG, "Missing account ID, beacon key, and/or server URL format");
+            return null;
+        }
 
-    private final ConfigProvider configProvider;
-    private final OkHttpClient client;
+        Log.d(TAG, String.format("Built HTTP request for batch of size %d: %s", batch.size(), batch.toString()));
```

Comment:
I realize this is a debug message, but are you sure you want the entire `batch` in here? It might get pretty verbose.

And nit: `%s` will do `toString()` for you.